File: Dynamic Programming/70. Climbing Stairs/70. Climbing Stairs.js
/**
 * @param {number} n
 * @return {number}
 */
var climbStairs = function (n) {
  if (n <= 2) return n;

  let prev2 = 1;
  let prev1 = 2;

  for (let i = 2; i < n; i++) {
    [prev2, prev1] = [prev1, prev2 + prev1];
  }

  return prev1;
};

// Time: O(n)
// Space: O(1)

File: Dynamic Programming/70. Climbing Stairs/70. Climbing Stairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        if n <= 2:
            return n
        
        prev2 = 1
        prev1 = 2

        for _ in range(2, n):
            prev2, prev1 = prev1, prev2 + prev1
        
        return prev1

# Time: O(n)
# Space: O(1)